compiler/sem/semexprs.nim

```diff
@@ -696,85 +696,135 @@ proc arrayConstrType(c: PContext, n: PNode): PType =
   typ[0] = makeRangeType(c, 0, n.len - 1, n.info)
   result = typ
 
+proc semArrayElement(c: PContext, n: PNode, formalIdx: PType
+                    ): tuple[n: PNode, idx: Int128, idxType: PType] =
+  ## Analyses the array element expression `n`, producing the *unchecked* value
+  ## expression plus the index value and type (if one is specified), or an
+  ## error. `formalIdx` specifies the type that the index value must fit. If
+  ## `formalIdx` is nil, the index specified by `n` is always allowed as long
+  ## as it's of a valid ordinal-like type.
+  ##
+  ## If an error occurs, the returned index type is nil.
+  addInNimDebugUtils(c.config, "semArrayElement", n, result.n)
+
+  case n.kind
+  of nkExprColonExpr:
+    checkSonsLen(n, 2, c.config)
+    let idx = semRealConstExpr(c, n[0])
+
+    if idx.kind == nkError:
+      # wrap in an error
+      let e = newNodeI(nkExprColonExpr, n.info, 2)
+      e[0] = idx
+      e[1] = n[1]
+      (c.config.wrapError(e), Zero, nil)
```

Comment:
Not a hard requirement, I just think it'd lead to a some simplifications. Here I'd return `idx` instead of wrapping the whole `nkExprColonExpr` in an error. This way one would only get: the index node, the index node error, or an empty node (changing the `else` branch).